S3 Unused Security Groups

horakj opened the pull request:
### Background

Unused Security Groups have the ability of being hazardous.  
Have the ability to make secondary boto3 calls gives the user the ability to create more in depth rules.

### Changes

* Added Global Function to create a Boto3 
* Added Unused Security Group Cloud Security Rule

### Testing

* Tested locally, Nick Hakmiller has PR for permissions adjustment for the policy engine. These two PRs rely on each other to work.

nhakmiller commented on aws_vpc_policies/aws_security_group_unused_security_group.yml:
```diff
@@ -0,0 +1,141 @@
+AnalysisType: policy
+Filename: aws_security_group_unused_security_group.py
+PolicyID: AWS.SecurityGroup.UnusedSecurityGroup
+Enabled: true
```
I would set this to `Enabled: false` by default, as this policy only works if a specific non-default deployment option is set.

nhakmiller commented on aws_vpc_policies/aws_security_group_unused_security_group.yml:
```diff
@@ -0,0 +1,141 @@
+  - AWS
+  - PCI
+Reports:
+  PCI:
+    - 1.1.4
+    - 1.2.1
+    - 8.2.1
```
Just double checking, are these PCI mappings correct? Just want to make sure they didn't end up here vvia copy/paste.

horakj replied:
Removed the PCI

nhakmiller commented on aws_vpc_policies/aws_security_group_unused_security_group.yml:
```diff
@@ -0,0 +1,141 @@
+Runbook: >
+  Delete unused Security Groups.
+Reference: https://docs.aws.amazon.com/cli/latest/userguide/cli-services-ec2-sg.html
+# Tests:
+#   -
+#     Name: Not In Use Security Group
+#     ExpectedResult: false
+#     Resource:
```
Probably best to just remove these tests if they're not being used.

nhakmiller commented on global_helpers/panther_oss_helpers.py:
```diff
@@ -16,6 +16,31 @@ class BadLookup(Exception):
 class PantherBadInput(Exception):
     """Error returned when a Panther helper function is provided bad input."""
 
+def build_client(resource):
+    """Function builds resource client that assumes Panther audit role"""
+    tmp = resource['ResourceId'].split(':')
+    resource_type = tmp[2]
+    region = tmp[3]
+    account = tmp[4]
+    role_arn = f'arn:aws:iam::{account}:role/PantherAuditRole-{region}'
+    sts_connection = boto3.client('sts')
+    acct_b = sts_connection.assume_role(
+        RoleArn=role_arn,
+        RoleSessionName="lambda_assume_audit_role"
+    )
```
Let's include the changes we discussed here, having the caller specify a service etc.

horakj replied:
Testing then sending the code changes we made.

horakj commented:
I am about to update the code with what we changed, did not get to it yesterday.

Best Regards,

Justin

> On Dec 9, 2020, at 8:45 AM, Nick <notifications@github.com> wrote:
> 
> ﻿
> @nhakmiller commented on this pull request.
> 
> In aws_vpc_policies/aws_security_group_unused_security_group.yml:
> 
> > @@ -0,0 +1,141 @@
> +AnalysisType: policy
> +Filename: aws_security_group_unused_security_group.py
> +PolicyID: AWS.SecurityGroup.UnusedSecurityGroup
> +Enabled: true
> I would set this to Enabled: false by default, as this policy only works if a specific non-default deployment option is set.
> 
> In aws_vpc_policies/aws_security_group_unused_security_group.yml:
> 
> > +  PCI:
> +    - 1.1.4
> +    - 1.2.1
> +    - 8.2.1
> Just double checking, are these PCI mappings correct? Just want to make sure they didn't end up here vvia copy/paste.
> 
> In aws_vpc_policies/aws_security_group_unused_security_group.yml:
> 
> > +# Tests:
> +#   -
> +#     Name: Not In Use Security Group
> +#     ExpectedResult: false
> +#     Resource:
> Probably best to just remove these tests if they're not being used.
> 
> In global_helpers/panther_oss_helpers.py:
> 
> > +def build_client(resource):
> +    """Function builds resource client that assumes Panther audit role"""
> +    tmp = resource['ResourceId'].split(':')
> +    resource_type = tmp[2]
> +    region = tmp[3]
> +    account = tmp[4]
> +    role_arn = f'arn:aws:iam::{account}:role/PantherAuditRole-{region}'
> +    sts_connection = boto3.client('sts')
> +    acct_b = sts_connection.assume_role(
> +        RoleArn=role_arn,
> +        RoleSessionName="lambda_assume_audit_role"
> +    )
> Let's include the changes we discussed here, having the caller specify a service etc.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

wey-chiang commented on aws_s3_policies/aws_s3_bucket_public_read.py:
```diff
@@ -6,9 +6,10 @@
-        if grant['Grantee']['URI'] in GRANTEES and grant[
-                'Permission'] in PERMISSIONS:
-            return False
+    if 'Grants' in resource and resource['Grants'] is not None:
```
Nit: the built-in `dict.get` should be used in place of `dict[key]` when possible

This applies to a lot of the python files in this PR

wey-chiang reviewed (changes requested):
Looks like `aws_s3_policies/aws_s3_bucket_public_write.yml` needs a newline at the EOF.

~You might need to run `make deps-update` to resolve the CircleCI build issues~
Edit: pulling in changes from the latest version of master should resolve these issues

jacknagz commented:
@horakj what's up with this PR?

horakj commented:
Last minute request for changes due to syntax that prevented the PR from moving forward, I will take care of it.

horakj commented:
Had to update branch with the master branch to make the the syntax changes requested by @wey-chiang which is why it look like a bunch of things going on.

wey-chiang commented on README.md:
````diff
@@ -1,268 +1,185 @@
 ```python
 def rule(event):
-  return False
+    return (event.get('outcome', {}).get('result', '') == 'FAILURE' and
+            event['eventType'] == 'user.session.start')
+
+
+def title(event):
+    return 'Suspected brute force Okta logins to account {} due to [{}]'.format(
+        event.get('actor', {}).get('alternateId', 'ID_NOT_PRESENT'),
+        event.get('outcome', {}).get('reason', 'REASON_NOT_PRESENT')
````
👍

wey-chiang commented on aws_cloudtrail_rules/aws_key_compromised.py:
```diff
@@ -3,8 +3,10 @@
     request_params = event.get('requestParameters', {})
-    return (event['eventName'] == 'PutUserPolicy' and
-            request_params.get('policyName') == EXPOSED_CRED_POLICY)
+    if request_params:
+        return (event['eventName'] == 'PutUserPolicy' and
+                request_params.get('policyName') == EXPOSED_CRED_POLICY)
+    return False
```
```suggestion
    if request_params:
        return (event.get('eventName', '') == 'PutUserPolicy' and
                request_params.get('policyName', '') == EXPOSED_CRED_POLICY)
    return False
```

wey-chiang commented on aws_cloudtrail_rules/aws_resource_made_public.py:
```diff
@@ -1,69 +1,78 @@
 # pylint: disable=too-many-return-statements
 def rule(event):
-    parameters = event.get('requestParameters', {})
+    if event.get('errorCode') == 'AccessDenied':
```
```suggestion
    if event.get('errorCode', '') == 'AccessDenied':
```

wey-chiang commented on aws_cloudtrail_rules/aws_resource_made_public.py:
```diff
@@ -1,69 +1,78 @@
     if event['eventName'] == 'PutBucketPolicy':
-        # Don't alert if access is denied
-        return policy_is_not_acceptable(parameters.get('bucketPolicy', None))
+        return policy_is_internet_accessible(parameters.get('bucketPolicy'))
```
```suggestion
        return policy_is_internet_accessible(parameters.get('bucketPolicy', None))
```

wey-chiang commented on aws_cloudtrail_rules/aws_resource_made_public.py:
```diff
@@ -1,69 +1,78 @@
-            policy = parameters.get('attributeValue', '{}')
-            return policy_is_not_acceptable(json.loads(policy))
-        return False
+    if event['eventName'] == 'SetTopicAttributes' and parameters.get(
+            'attributeName') == 'Policy':
```
```suggestion
    if event['eventName'] == 'SetTopicAttributes' and parameters.get(
            'attributeName', '') == 'Policy':
```

wey-chiang commented on aws_cloudtrail_rules/aws_resource_made_public.py:
```diff
@@ -1,69 +1,78 @@
+        event, 'userIdentity', 'sessionContext', 'sessionIssuer', 'userName')
+
+    if event.get('Resources'):
+        return f"AWS Resource {event.get('Resources')[0]['arn']} made public by {user}"
```
```suggestion
        return f"AWS Resource {event.get('Resources')[0].get('arn', 'MISSING ARN')} made public by {user}"
```

wey-chiang commented on aws_cloudtrail_rules/aws_resource_made_public.py:
```diff
@@ -1,69 +1,78 @@
+    if event.get('Resources'):
+        return f"AWS Resource {event.get('Resources')[0]['arn']} made public by {user}"
+
+    return f"{event['eventSource']} resource made public by {user}"
```
```suggestion
    return f"{event.get('eventSource', 'MISSING SOURCE')} resource made public by {user}"
```

wey-chiang commented on aws_vpc_policies/aws_security_group_unused_security_group.py:
```diff
@@ -0,0 +1,10 @@
+def policy(resource):    
+    client = build_client(resource, 'ec2', resource['Region'])
+    results = client.describe_network_interfaces(Filters=[{'Name' : 'group-id', 'Values' : [resource['Id']]}])
+    if results['NetworkInterfaces']:
```
```suggestion
    if results.get('NetworkInterfaces'):
```

wey-chiang commented on global_helpers/panther_audit.py:
```diff
@@ -0,0 +1,35 @@
+        RoleArn=role_arn,
+        RoleSessionName="lambda_assume_audit_role"
+    )
+    ACCESS_KEY = acct_b['Credentials']['AccessKeyId']
+    SECRET_KEY = acct_b['Credentials']['SecretAccessKey']
+    SESSION_TOKEN = acct_b['Credentials']['SessionToken']
```
```suggestion
    ACCESS_KEY = acct_b.get('Credentials', {}).get('AccessKeyId', 'MISSING ACCESS KEY')
    SECRET_KEY = acct_b.get('Credentials', {}).get('SecretAccessKey', 'MISSING SECRET ACCESS KEY')
    SESSION_TOKEN = acct_b.get('Credentials', {}).get('SessionToken', 'MISSING SESSION TOKEN')
```

jacknagz replied:
Use DeepGet instead of multiple gets!!!

https://github.com/panther-labs/panther-analysis/blob/master/global_helpers/panther_base_helpers.py#L147-L154

horakj replied:
This is for my panther_audit global module those creds come from the boto3 sts assume role call, there is really no need to use deep dive or gets. If the call fails then we will never get to the part of assigning the keys and tokens. This part is straight from AWS Docs, I can change it, but I don't think its needed.

jacknagz replied:
I'm going to say this when I see multiple .get calls

wey-chiang commented on okta_rules/okta_admin_role_assigned.py:
```diff
@@ -1,27 +1,44 @@
-            re.search(
-                r'[aA]dministrator',
-                event['debugContext']['debugData'].get('privilegeGranted'))))
+    return (event['eventType'] == 'user.account.privilege.grant' and
```
```suggestion
    return (event.get('eventType', None) == 'user.account.privilege.grant' and
```

jacknagz replied:
why use .get on a required field?

wey-chiang commented on okta_rules/okta_admin_role_assigned.py:
```diff
@@ -1,27 +1,44 @@
 def title(event):
     title_str = '{} <{}> granted [{}] privileges to {} <{}>'
+
+    target = event.get('target', [])
+    display_name = target[0]['displayName'] if target else ''
+    alternate_id = target[0]['alternateId'] if target else ''
```
```suggestion
    target = event.get('target', [{}])
    display_name = target[0].get('displayName', 'MISSING DISPLAY NAME') if target else ''
    alternate_id = target[0].get('alternateId', 'MISSING ALTERNATE ID') if target else ''
```

wey-chiang reviewed (approved):
Nice, looks really good! 

@austinbyers your use of github's suggest code changes has inspired me!

Made a few suggestions here and there for things that wouldn't intentionally change functionality, but provide some more resilience 😄
